src/main/scala/diplomacy/Nodes.scala

```diff
@@ -5,123 +5,328 @@ package freechips.rocketchip.diplomacy
+    * @param sourceInfo [[SourceInfo]] of this edge
+    * @return an outward edge of this node
+    */
   def edgeO(pd: DO, pu: UO, p: Parameters, sourceInfo: SourceInfo): EO
+
+  /** Create an outward Bundle parameterized by the inward edge
+    *
+    * @param eo the outward Edge of this node
+    * @return an outward Bundle of this node
+    */
   def bundleO(eo: EO): BO
 
-  // optional methods to track node graph
-  def mixO(pd: DO, node: OutwardNode[DO, UO, BO]): DO = pd // insert node into parameters
-  def getI(pd: DO): Option[BaseNode] = None // most-inward common node
+  /** Defines how outward parameters can be "mixed" or negotiated together.
+    *
+    * The default behavior is to just return `pd`.
+    *
+    * @param pd   downward flowing edge parameters
+    * @param node an outward node to mix the downward flowing parameters into
+    * @return altered version of the downward flowing outward parameters
+    */
+  def mixO(pd: DO, node: OutwardNode[DO, UO, BO]): DO = pd
+
+  /** Return the most-inward common node */
+  def getI(pd: DO): Option[BaseNode] = None
 }
 
+/** The [[NodeImp]] combines an [[InwardNodeImp]] and an [[OutwardNodeImp]].
+  *
+  * This allows it to define whether it is a protocol-modifying (bridging)
+  * sort of node, or whether it is an adapter type node that just modifies
+  * the parameters within a protocol.
+  *
+  * This class has no members and is solely used for holding type information.
+  * Applications of diplomacy should extend [[NodeImp]] with a case object that sets
+  * concrete type arguments.
+  *
+  * @tparam D  type of the Downwards flowing parameters of the node
+  * @tparam U  type of Upwards flowing parameters of the node
+  * @tparam EO type of the parameters describing an edge on the outer side of the node
+  * @tparam EI type of the parameters describing an edge on the inner side of the node
+  * @tparam B  Bundle type generated by connections to this node
+  */
 abstract class NodeImp[D, U, EO, EI, B <: Data]
   extends Object with InwardNodeImp[D, U, EI, B] with OutwardNodeImp[D, U, EO, B]
 
-// If your edges have the same direction, using this saves you some typing
+/** A [[NodeImp]] where the inward and outward edges are of the same type.
+  *
+  * If the parameters on inward flowing edges are the same as parameters on outer flowing edges, the [[SimpleNodeImp]] can be used.
+  *
+  * @tparam D type of the Downwards flowing parameters of the node
+  * @tparam U type of the Upwards flowing parameters of the node
+  * @tparam E Edge Parameters describing the connections on either side of the node
+  * @tparam B Bundle type generated on edges connecting to this node.
+  *
+  *           If your edges have the same type in both directions, using this saves you some typing.
+  */
 abstract class SimpleNodeImp[D, U, E, B <: Data]
   extends NodeImp[D, U, E, E, B]
 {
+  /** Create an edge that that passes through this node.
```

Comment:
Edges do not "pass through nodes".

edge creates the edge parameters out of the downward flowing and upward flowing parameters for edges that connect to this node.